int32 velocity_x
int32 velocity_y